dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs

```diff
@@ -0,0 +1,274 @@
+            }
+
+            if (!this._inputs.TryGetValue(message.FunctionName, out Dictionary<string, object?>? functionParameters))
+            {
+                this._inputs[message.FunctionName] = new();
+                functionParameters = this._inputs[message.FunctionName];
+            }
+
+            functionParameters![kvp.Key] = kvp.Value;
+        }
+
+        // If we're still waiting for inputs on all of our functions then don't do anything.
+        List<string> invocableFunctions = this._inputs.Where(i => i.Value != null && i.Value.All(v => v.Value != null)).Select(i => i.Key).ToList();
+        var missingKeys = this._inputs.Where(i => i.Value is null || i.Value.Any(v => v.Value is null));
+
+        if (invocableFunctions.Count == 0)
+        {
+            string missingKeysLog() => string.Join(", ", missingKeys.Select(k => $"{k.Key}: {string.Join(", ", k.Value?.Where(v => v.Value == null).Select(v => v.Key) ?? [])}"));
+            this._logger?.LogDebug("No invocable functions, missing keys: {MissingKeys}", missingKeysLog());
+            return;
+        }
+
+        // A message can only target one function and should not result in a different function being invoked.
+        var targetFunction = invocableFunctions.FirstOrDefault((name) => name == message.FunctionName) ??
+            throw new InvalidOperationException($"A message targeting function '{message.FunctionName}' has resulted in a function named '{invocableFunctions.First()}' becoming invocable. Are the function names configured correctly?");
+
+        this._logger?.LogDebug("Step with Id `{StepId}` received all required input for function [{TargetFunction}] and is executing.", this.Name, targetFunction);
+
+        // Concat all the inputs and run the function
+        KernelArguments arguments = new(this._inputs[targetFunction]!);
+        if (!this._functions.TryGetValue(targetFunction, out KernelFunction? function) || function == null)
+        {
+            throw new ArgumentException($"Function {targetFunction} not found in plugin {this.Name}");
+        }
+
+        FunctionResult? invokeResult = null;
+        string? eventName = null;
+        object? eventValue = null;
+
+        // Invoke the function, catching all exceptions that it may throw, and then post the appropriate event.
+#pragma warning disable CA1031 // Do not catch general exception types
+        try
+        {
+            invokeResult = await this.InvokeFunction(function, this._kernel, arguments).ConfigureAwait(false);
+            eventName = $"{targetFunction}.OnResult";
+            eventValue = invokeResult?.GetValue<object>();
+        }
+        catch (Exception ex)
+        {
+            this._logger?.LogError("Error in Step {StepName}: {ErrorMessage}", this.Name, ex.Message);
```

Comment:
Can we please consider filing an issue to introduce schematized logging messages.  Going to be a lot easier to start there than refactor a bunch of logging later.

https://learn.microsoft.com/en-us/dotnet/core/extensions/logger-message-generator